lib/backup_restore/s3_backup_store.rb

```diff
@@ -0,0 +1,90 @@
+require_dependency "backup_restore/backup_store"
+require_dependency "s3_helper"
+
+module BackupRestore
+  class S3BackupStore < BackupStore
+    DOWNLOAD_URL_EXPIRES_AFTER_SECONDS = 15
+    UPLOAD_URL_EXPIRES_AFTER_SECONDS = 21600 # 6 hours
+
+    def initialize(opts = {})
+      s3_options = S3Helper.s3_options(SiteSetting)
+      s3_options.merge!(opts[:s3_options]) if opts[:s3_options]
+      @s3_helper = S3Helper.new(SiteSetting.s3_backup_bucket, '', s3_options)
+    end
+
+    def remote?
+      true
+    end
+
+    def file(filename, include_download_source: false)
+      obj = @s3_helper.object(filename)
+      create_file_from_object(obj, include_download_source) if obj.exists?
+    end
+
+    def delete_file(filename)
+      obj = @s3_helper.object(filename)
+      obj.delete if obj.exists?
+    end
+
+    def download_file(filename, destination_path, failure_message = nil)
+      unless @s3_helper.object(filename).download_file(destination_path)
+        raise failure_message.presence&.to_s || "Failed to download file"
+      end
+    end
+
+    def upload_file(filename, source_path, content_type)
+      obj = @s3_helper.object(filename)
+      raise BackupFileExists.new if obj.exists?
+
+      obj.upload_file(source_path, content_type: content_type)
+    end
+
+    def generate_upload_url(filename)
+      obj = @s3_helper.object(filename)
+      raise BackupFileExists.new if obj.exists?
+
+      presigned_url(obj, :put, UPLOAD_URL_EXPIRES_AFTER_SECONDS)
+    end
+
+    private
+
+    def unsorted_files
+      @s3_helper
+        .list
+        .select { |obj| obj.key.match?(/\.t?gz$/i) }
```

Comment:
Just curious: would it be possible to do the filtering using the S3 API?

Reply:
Unfortunately that's not possible. The API allows only filtering by prefix.

Reply:
Let's use `select!` here in case the list is large.

Reply:
`list` returns up to 1000 elements. `Aws::S3::ObjectSummary::Collection` doesn't support `select!`. I could convert it to an array, but I guess that's even more inefficient than using `select`. I'd say in 99% of all cases `select` won't remove any elements, so it should be fine.